### ISSUE TITLE
no-cycle: allowUnsafeDynamicCyclicDependency also suppresses fully static cycles

### ISSUE THREAD

**Issue by wimvanschandevijl** [un-ts/eslint-plugin-import-x#515]
### Summary

`allowUnsafeDynamicCyclicDependency` is documented as *"Allow cyclic dependency if there is at least one dynamic import in the chain"*. In practice it also suppresses cycles that contain **no dynamic import at all**, because the check aborts traversal of a module's *entire* import list rather than skipping the dynamic edge.

The result is a silent false negative: enabling the option to ignore `lazy(() => import(...))` boundaries also stops reporting genuine static cycles that merely pass through a module which happens to have a dynamic import somewhere in it.

### Versions

- `eslint-plugin-import-x` 4.17.1 (latest at time of writing)
- `eslint` 9.39.5
- Node 24

### Reproduction

Three modules. `a` and `c` form a **purely static** cycle. `b` is only ever dynamically imported by `a`, and is not part of any cycle.

```js
// a.js
import { c } from './c.js';
export const loadB = () => import('./b.js');   // dynamic; b is NOT part of any cycle
export const a = () => c();
```

```js
// b.js
export const b = 1;
```

```js
// c.js
import { a } from './a.js';
export const c = () => a;
```

```js
// eslint.config.mjs
import importX from 'eslint-plugin-import-x';
export default [
  {
    files: ['**/*.js'],
    plugins: { 'import-x': importX },
    rules: { 'import-x/no-cycle': ['error'] },        // vs. ['error', { allowUnsafeDynamicCyclicDependency: true }]
  },
];
```

**Default (option off)** — both ends of the static `a ↔ c` cycle are reported, as expected:

```
a.js:1  Dependency cycle detected
c.js:1  Dependency cycle detected
```

**With `allowUnsafeDynamicCyclicDependency: true`:**

```
a.js:1  Dependency cycle detected
```

`c.js` is gone. Nothing about the `a ↔ c` cycle is dynamic — the only dynamic import in the project points at `b`, which is not in the cycle. Deleting `loadB` from `a.js` brings the `c.js` report back.

### Cause

`src/rules/no-cycle.ts` (built output at `lib/rules/no-cycle.js:90-105`):

```js
for (const [path, { getter, declarations }] of m.imports) {
  if (traversed.has(path)) {
    continue;
  }
  const toTraverse = [...declarations].filter(({ source, isOnlyImportingTypes }) =>
    !ignoreModule(source.value) && !isOnlyImportingTypes);

  if (options.allowUnsafeDynamicCyclicDependency && toTraverse.some(d => d.dynamic)) {
    return;                       // <-- exits detectCycle entirely
  }

  if (path === filename && toTraverse.length > 0) {
    return true;
  }
  ...
}
```

The guard is inside the loop over `m.imports`, but `return` exits `detectCycle` rather than continuing to the next import. So as soon as any one of the module's imports is dynamic, the module's **remaining imports are never examined** — including static ones that would close a real cycle. `traversed.add(m.path)` has already run above the loop, so the module is not revisited from another route either.

Whether a given static cycle survives therefore depends on where the dynamic edge falls in the iteration order of `m.imports`, which is not something the user can reason about.

### Expected

The option should skip the *dynamic edge*, not abandon the module. I would expect `continue` rather than `return`:

```js
  if (options.allowUnsafeDynamicCyclicDependency && toTraverse.some(d => d.dynamic)) {
    continue;
  }
```

That matches the documented description ("allow cyclic dependency **if there is at least one dynamic import in the chain**" — i.e. the chain through the dynamic edge), and keeps cycles composed only of static edges reported.

A stricter variant would filter out just the dynamic declarations and keep traversing the static ones for the same path, since a path can be imported both statically and dynamically:

```js
  const toTraverse = [...declarations].filter(({ source, isOnlyImportingTypes, dynamic }) =>
    !ignoreModule(source.value) && !isOnlyImportingTypes &&
    !(options.allowUnsafeDynamicCyclicDependency && dynamic));
  if (toTraverse.length === 0) continue;
```

### Why it matters in practice

I hit this evaluating the option on a production React app (~660 modules, 63 reported cycles). Enabling it removed 28 reports. 24 of those genuinely closed only through a `React.lazy(() => import(...))` boundary — the intended effect. The other **4 were fully static cycles**, verified by building a static-only import graph (`import type` and every `import()` excluded) and confirming a static path still ran from the imported module back to the importer.

One of them was a nine-module loop that ended up with no report anywhere in the project, purely because one module on the path uses `lazy()` for an unrelated component. Since the suppressions documenting those cycles also become "unused directives" under `--report-unused-disable-directives`, the option and the documentation for those cycles are mutually exclusive — there is no way to enable it and keep the static cycles flagged.

Happy to open a PR with the `continue` change plus a regression test along the lines of the repro above, if you agree with the direction.

**Comment by JounQin** [comment]
PR welcome for regression test cases and fixes.
